### PR TITLE
Delete trunk on failure

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

part of #1085 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by jichenjc** [comment]
@mdbooth I steal your code https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1087/files :)

File: pkg/cloud/services/compute/instance.go [hunk 1]
```diff
@@ -489,7 +489,17 @@ func (s *Service) DeleteInstance(eventObject runtime.Object, instance *InstanceS
 }
 
 func (s *Service) deletePorts(eventObject runtime.Object, nets []servers.Network) error {
+	trunkSupported, err := s.isTrunkExtSupported()
+	if err != nil {
+		return fmt.Errorf("obtaining network extensions: %v", err)
```
Comment (seanschneeweiss) [inline pkg/cloud/services/compute/instance.go:494]: I think that is ok to just return the err here, as it is already wrapped with `there was an issue verifying whether trunk support is available`.

**Review by seanschneeweiss — COMMENTED** [review]
/lgtm

File: pkg/cloud/services/compute/instance.go (hunk 1 above)
Comment (jichenjc) [reply pkg/cloud/services/compute/instance.go:494]: ok

**Comment by seanschneeweiss** [comment]
looks good, thank you
/lgtm

**Comment by jichenjc** [comment]
@mdbooth could you please help to take a look? Thanks

**Review by mdbooth — APPROVED** [review]
/lgtm

Thanks!

Incidentally, I've mentioned this before, but if we ever have the time I'd like to split resource creation out into separate reconcile functions which would make this code simpler to understand and test. The size of a single CreateInstance unit test feels like a warning.

I probably should have started this with my Cinder-AZ patch (i.e. have a separate reconcile loop for volumes). The deletion logic is over-complex there, too. However, now it's written I'd really like to get that landed so we can use the new API.

**Comment by jichenjc** [comment]
/hold cancel

ok, agree and let's create new issue (or reuse existing one? ) to track the TODOs you mentioned..

**Comment by mdbooth** [comment]
> @jichenjc: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-provider-openstack-e2e-test	[28f820b](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/28f820b3580ac901864b169560ab89b325f323bd)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1095/pull-cluster-api-provider-openstack-e2e-test/1480848100384837632)	true	`/test pull-cluster-api-provider-openstack-e2e-test`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-openstack&pr=1095). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Ajichenjc). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues?q=is:issue+is:open) when you hit one in your PR.

```
+ gcloud compute disks describe capo-e2e-disk --project k8s-jkns-gci-gce-alpha --zone us-east4-a
ERROR: (gcloud.compute.disks.describe) HTTPError 404: The resource 'projects/k8s-jkns-gci-gce-alpha/zones/us-east4-a/disks/capo-e2e-disk' was not found
+ gcloud compute disks create capo-e2e-disk --project k8s-jkns-gci-gce-alpha --image-project ubuntu-os-cloud --image-family ubuntu-2004-lts --zone us-east4-a
ERROR: (gcloud.compute.disks.create) Could not fetch resource:
 - The resource 'projects/ubuntu-os-cloud/zones/us-east4-a/imageFamilyViews/ubuntu-2004-lts' was not found
```

Did we get unlucky and that resource was being refreshed, or is something down or permanently changed?

**Pull request merged**